Adopt `parsers` in Filebeat's journald input

belimawr opened the pull request:
Splits from #26130

<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

Adopt `parsers` in Filebeat's journald input

## Why is it important?

It enables Filebeat to parse more types of log from journald input

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [x] Integration tests using a real journald

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Relates #26130

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29070/detail/PR-29070/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29070/detail/PR-29070/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29070/detail/PR-29070/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29070/detail/PR-29070/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_29070.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2021-11-22T09:50:28.212Z&rangeTo=2021-11-22T10:10:28.212Z&transactionName=Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=unknown&latencyAggregationType=avg&traceId=5df5416f4b0d899a94ec56270320c526&transactionId=5dc5557e7d411a58)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2021-11-22T10:00:28.212+0000


* Duration: 14 min 8 sec


* Commit: efff3922151951f79884dbb9a89058eb123424bd



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3  |
| Skipped      | 0 |
| Total        | 3   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->











### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29070/detail/PR-29070/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `filebeat-lint - make -C filebeat check;make -C filebeat update;make -C x-pack/filebeat check;make`
<ul>
<li>Took 2 min 56 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-29070/runs/1/steps/1349/log/?start=0">here</a></li>
<li>Description: <code>make -C filebeat check;make -C filebeat update;make -C x-pack/filebeat check;make -C x-pack/filebeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `x-pack/filebeat-lint - make -C x-pack/filebeat check;make -C x-pack/filebeat update;make -C filebe`
<ul>
<li>Took 2 min 58 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-29070/runs/1/steps/1355/log/?start=0">here</a></li>
<li>Description: <code>make -C x-pack/filebeat check;make -C x-pack/filebeat update;make -C filebeat check;make -C filebeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-29070/runs/1/steps/1358/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 2'</code></l1>
</ul>
  
  </p>
  </details>


## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

kvch commented on filebeat/input/journald/input.go:
```diff
@@ -204,3 +218,45 @@ func seekBy(log *logp.Logger, cp checkpoint, seek, defaultSeek journalread.SeekM
+	return r.r.Close()
+}
+
+func (r *readerAdapter) Next() (reader.Message, error) {
```
`Meta` is a legacy attribute we would like to get rid of. So please avoid using it.

Instead of putting partial state information to `Meta`, you should add a `Private` attribute to `Message` and save the state/checkpoint information there in this `Next` function.

Example

```golang

func (r *readerAdapter) Next() (reader.Message, error) {
   // ...
    m := reader.Message{
        Private: checkpoint{
            Version: cursorVersion,
            RealtimeTimestamp: data.RealtimeTimestamp,
            MonotonicTimestamp: data.MonotonicTimestamp,
            Position: data.Cursor,
            Path: r.path,
      }
    }
```

Then it can be published like this:
```golang
if err := publisher.Publish(event, event.Private); err != nil {
```

belimawr replied:
Done, I force pushed the changes alongside other improvements. I hope I got it right.

belimawr commented on filebeat/_meta/config/filebeat.inputs.reference.yml.tmpl:
```diff
@@ -557,3 +557,12 @@ filebeat.inputs:
   # Configure stream to filter to a specific stream: stdout, stderr or all (default)
   #stream: all
 
+#------------------------------ Journald input --------------------------------
```
@kvch  Does this config need to be added here? Journald input is still experimental and before there was no mention to it in the reference configuration, so I'm not sure whether it should be added or not.

kvch replied:
Yes, it's ok if you add it. Please add a sentence about the input, including that is experimental.

belimawr replied:
Done! (changes will be pushed soon).

Quick questions:
1. Do I need to add the full set of configuration? 
2. Should I link somehow to the [documentation](https://www.elastic.co/guide/en/beats/filebeat/master/filebeat-input-journald.html)

I feel that compared to all other inputs journald is missing a proper documentation in the reference config. Or is it like that because it's experimental?

kvch replied:
It is good as it is. I have planned a few breaking changes in filtering. When that is implemented, I will write a more detailed documentation page for it.

belimawr commented on filebeat/filebeat_parsers_test.go:
```diff
@@ -0,0 +1,132 @@
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build linux && cgo && withjournald
```
@kvch do I need to add the `integration` flag here?

kvch replied:
We do not need such a high-level test for journald input in beats. I would rather we put such tests in `e2e-testing` repository. For testing parsers in journald input, you should check out how it is done for the filestream input: https://github.com/elastic/beats/blob/master/filebeat/input/filestream/parsers_integration_test.go

Regardless, I would get this PR in without this test, and add integration tests in a follow up PR. WDYT?

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane  (Team:Elastic-Agent-Data-Plane)

kvch commented on filebeat/input/journald/input.go:
```diff
@@ -204,3 +205,44 @@ func seekBy(log *logp.Logger, cp checkpoint, seek, defaultSeek journalread.SeekM
+	}
+
+	content := []byte(data.Fields["MESSAGE"])
+	delete(data.Fields, "MESSAGE")
```
```suggestion
	err = data.Fields.Delete("MESSAGE")
	if err != nil {
	    return reader.Message{}, fmt.Errorf("error deleting MESSAGE key: %+v", err)
	}
```

belimawr replied:
Unfortunately, that won't work, `data.Fields` is a `map[string]string`, hence it hasn't got a `Delete` method. `data` is a `sdjournal.JournalEntry` (from `github.com/coreos/go-systemd/v22/sdjournal`).

kvch replied:
Oh yes, my bad. I thought it `r.r.Next` returns `reader.Message` not a `journalread.Entry`. :facepalm:

kvch reviewed (changes requested):
Let's remove the e2e test and add a simpler integration test in a follow up PR.

belimawr commented:
> Let's remove the e2e test and add a simpler integration test in a follow up PR.

I removed it and added a unit test on: filebeat/input/journald/input_parsers_test.go

kvch commented on filebeat/input/journald/environment_test.go:
```diff
@@ -0,0 +1,306 @@
+	grp unison.TaskGroup
+}
+
+//TODO(Tiago): clean up un-used suff like the temp dir
```
Please remove this line.

BTW, I am not sure what you mean by this line. But if you refer to the directory created by `t.TempDir`, it is automatically cleaned up during the tear down phase of the test.

belimawr replied:
That comment does not belong there, it was from a previous iteration :sweat_smile: 

Thanks for catching it!

kvch commented on filebeat/input/journald/environment_test.go:
```diff
@@ -0,0 +1,306 @@
+	}
+}
+
+// requireOffsetInRegistry checks if the expected offset is set for a file.
```
Please remove this.

belimawr replied:
Done!

belimawr commented:
rebase onto master, force push
